--- ./src/3rdparty/webkit/Source/WebCore/platform/graphics/qt/FontPlatformDataQt.cpp.orig
+++ ./src/3rdparty/webkit/Source/WebCore/platform/graphics/qt/FontPlatformDataQt.cpp	
@@ -45,7 +45,8 @@
     QFont& font = m_data->font;
     int requestedSize = qRound(description.computedPixelSize());
     font.setFamily(familyName);
-    font.setPixelSize(qRound(requestedSize));
+    if (requestedSize)
+    	font.setPixelSize(qRound(requestedSize));
     font.setItalic(description.italic());
     font.setWeight(toQFontWeight(description.weight()));
     font.setWordSpacing(wordSpacing);